Multi permit

ghalliday opened the pull request:
<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [ ] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

richardkchapman commented on roxie/udplib/udpsha.cpp:
```diff
@@ -34,18 +34,38 @@ using roxiemem::IDataBufferManager;
+
+//The following control the timeouts within the udp layer.  All timings are in milliseconds, but I suspect some of these should possibly be sub-millisecond
+unsigned udpFlowAckTimeout = 2;         // the maximum time that it is expected to take to receive an acknowledgement of a flow message (when one is sent) - should be small
+unsigned updDataSendTimeout = 20;       // how long to receive the maximum amount of data, ~100 packets of 8K should take 10ms on a 1Gb network. Timeout for assuming send_complete has been lost
```
Note: The comment suggests that this timeout needs to be dependent on the queue size (or the max slots per client) and possibly the MTU size in use.

richardkchapman commented on roxie/udplib/udpsha.cpp:
```diff
@@ -34,18 +34,38 @@ using roxiemem::IDataBufferManager;
+                                        // 0 means they are unlikely to be lost, so worth resending as soon as it appears to be missing - trading duplicate packets for delays
+
+unsigned udpMaxPendingPermits = 1;      // default to 5?
+bool udpResendAllMissingPackets = false;  // Should probably default to true
```
Dubious about what this means, whether it defeats the overlapping of multiple bursts from a single sender, and what impact of latency on throughput might be if we put request_to_send_more on the data socket.

richardkchapman commented on roxie/udplib/udpsha.cpp:
```diff
@@ -34,18 +34,38 @@ using roxiemem::IDataBufferManager;
+bool udpResendAllMissingPackets = false;  // Should probably default to true
+bool udpResendLostPackets = true;
+bool udpAssumeSequential = false;
+bool udpAdjustThreadPriorities = false;
```
This is a significant change to current behaviour...

richardkchapman commented on roxie/udplib/udpsha.cpp:
```diff
@@ -293,6 +313,9 @@ int check_max_socket_write_buffer(int size) {
 #if defined( __linux__) || defined(__APPLE__)
 void setLinuxThreadPriority(int level)
 {
+    if (!udpAdjustThreadPriorities)
```
What about windows?

richardkchapman commented on roxie/udplib/udpsha.hpp:
```diff
@@ -202,14 +202,29 @@ class simple_queue
+
+    // The following messages
+        request_received,       // acknowledge request to send from the sender
+        send_start,             // about to send data - indicate how many packets are actually going to be send
```
typo: going to be send

richardkchapman commented on roxie/udplib/udpsha.hpp:
```diff
@@ -202,14 +202,29 @@ class simple_queue
+    // The following messages
+        request_received,       // acknowledge request to send from the sender
+        send_start,             // about to send data - indicate how many packets are actually going to be send
+        request_to_send_more,   // equivalent to send_completed followed by a request_to_send
```
May want to uncombine these if one should go on data socket (so as not to overtake associated data) but other should not.

richardkchapman commented on roxie/udplib/udpsim.cpp:
```diff
@@ -132,16 +137,23 @@ void initOptions(int argc, const char **argv)
     udpDropFlowPackets[flowType::request_to_send_more] = options->getPropInt("@dropRequestToSendMorePackets", 0);  // drop 1 in N
-    udpDropFlowPackets[flowType::send_completed] = options->getPropInt("@dropSendCompletedPackets", 0);  // drop 1 in N
+    udpDropFlowPackets[flowType::send_start] = options->getPropInt("@dropSendStartPackets", 0);  // drop 1 in N
+    udpDropFlowPackets[flowType::send_completed] = options->getPropInt("@   ", 0);  // drop 1 in N
```
bug: code looks corrupt here?

richardkchapman commented on roxie/udplib/udpsim.cpp:
```diff
@@ -150,9 +162,10 @@ void initOptions(int argc, const char **argv)
-    constexpr unsigned maxSendQueueSize = 100;
+    //If a single permit allowed then grant that client all permits, otherwise allocate a share of twice the potential number of slots
+    const unsigned numReceiveSlots = 100;
+    const unsigned maxSlotsPerClient = (udpMaxPendingPermits == 1) ? numReceiveSlots : (2 * numReceiveSlots) / udpMaxPendingPermits;
```
Should this calculation be in send manager?

richardkchapman commented on roxie/udplib/udpsim.cpp:
```diff
@@ -164,7 +177,7 @@ void simulateTraffic()
             IpAddress pretendIP(VStringBuffer("8.8.8.%d", i));
             // Note - this is assuming we send flow on the data port (that option defaults true in roxie too)
-            Owned<ISendManager> sm = createSendManager(CCD_DATA_PORT, CCD_DATA_PORT, CCD_CLIENT_FLOW_PORT, maxSendQueueSize, 3, pretendIP, nullptr, false);
+            Owned<ISendManager> sm = createSendManager(CCD_SERVER_FLOW_PORT, CCD_DATA_PORT, CCD_CLIENT_FLOW_PORT, maxSendQueueSize, 3, pretendIP, nullptr, false);
```
Should probably be controlled by an option

richardkchapman commented on roxie/udplib/udptrr.cpp:
```diff
@@ -433,20 +671,171 @@ class CReceiveManager : implements IReceiveManager, public CInterface
+
+            //If the flowSeq is lower than the previous flowSeq there could be two causes:
+            //a) The sender has restarted
+            //b) Messages have been received out of order (e.g. equest_to_send_more after a request_to_send?)
```
typo: request

richardkchapman commented on roxie/udplib/udptrr.cpp:
```diff
@@ -433,20 +671,171 @@ class CReceiveManager : implements IReceiveManager, public CInterface
+            //=> ignore the flow if it was the previous seq, and could not come from the server restarting
+            unsigned flowSeq = msg.flowSeq;
+            const unsigned startupFlowSeqs = 2; // flow seq below this could be from a restarting server
+            if ((flowSeq + 1 == sender->flowSeq) && (flowSeq >= startupFlowSeqs))
```
Can you never be more than one ahead?

ghalliday replied:
I think so, but needs to be checked, and rethought if completed flow messages are sent on the flow thread.

richardkchapman commented on roxie/udplib/udptrr.cpp:
```diff
@@ -433,20 +671,171 @@ class CReceiveManager : implements IReceiveManager, public CInterface
+                        else
+                        {
+                            // Put it back on the queue (at the back)
+                            finger->permitTime = 0;
```
bug: I don't understand why 0 (rather than now) would be wanted here

richardkchapman replied:
Ok, then perhaps would be clearer/better to move this line to just after pendingPermits.remove(finger) ?

richardkchapman replied:
(and set finger->activePermits to 0)

richardkchapman commented on roxie/udplib/udptrr.cpp:
```diff
@@ -433,20 +671,171 @@ class CReceiveManager : implements IReceiveManager, public CInterface
+                // Sum any remaining permits
+                while (finger)
+                {
+                    permitsIssued += finger->getActivePermits();
```
Shouldn't we have accumulated these inside the loop above as well? I don't see where we do so.

richardkchapman replied:
ok

richardkchapman reviewed (changes requested):
First scan completed, need to look more later

richardkchapman commented on roxie/udplib/udptrr.cpp:
```diff
@@ -171,23 +610,24 @@ class CReceiveManager : implements IReceiveManager, public CInterface
-                return true;
-            return false;
+
+            //The best approximation to the oldest-inlight packet - because permits may have expired...
```
typo: inlight

richardkchapman commented on roxie/udplib/udptrr.cpp:
```diff
@@ -94,6 +492,43 @@ class CReceiveManager : implements IReceiveManager, public CInterface
+
+    //Increasing this number, increases the number of concurrent permits a sender may have (and its resilience to lost flow messages),
+    //but also increases the processing cost since code often iterates through all the permits.  2..4 likely to be good values.
+    static constexpr unsigned MAX_PERMITS = 2;
```
Might be better to call it MAX_PERMITS_PER_SENDER

richardkchapman commented on roxie/udplib/udptrr.cpp:
```diff
@@ -270,71 +710,118 @@ class CReceiveManager : implements IReceiveManager, public CInterface
-        UdpSenderEntry *head = nullptr;
-        UdpSenderEntry *tail = nullptr;
-        unsigned numEntries = 0;
+        //Total permits outstanding for the sender
```
Total reservations, I think. There is scope for confusion when using the term 'permit' between an active permission to send and the number of slots that the permit grants...

richardkchapman commented on roxie/udplib/udptrr.cpp:
```diff
@@ -270,71 +710,118 @@ class CReceiveManager : implements IReceiveManager, public CInterface
+        }
+        inline void decPermit(unsigned msgSeq)
+        {
+            //Although this is a bit more work than matching by flowSeq it shouldn't be too inefficient.
```
Usually the permit will be the same one that we matched last time. Can we do a quick check to tell?

richardkchapman replied:
This code feels painful to be executing per received message

richardkchapman commented on roxie/udplib/udptrr.cpp:
```diff
@@ -433,20 +971,195 @@ class CReceiveManager : implements IReceiveManager, public CInterface
+            }
+
+            // Aim is to issue enough permits to use all available the space in the queue.
+            // As data is received the number of slots reduces, but the relevant permit is reduced to match
```
I don't really understand this comment

richardkchapman commented on roxie/udplib/udptrr.cpp:
```diff
@@ -433,20 +971,195 @@ class CReceiveManager : implements IReceiveManager, public CInterface
+            bool anyCannotSend = false;
+            for (UdpSenderEntry *finger = pendingRequests; finger != nullptr; finger = finger->next)
+            {
+                //MORE: not sure if the permitsIssued test is required
```
MORE: should work it out then!

richardkchapman reviewed (commented):
General comment: this is beyond the scope of 8.4.x point release - but there may be some changes within it that need to go in 8.4 ?

ghalliday committed:
HPCC-26813 Refactor UDP flow control to allow multiple permits

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

ghalliday committed:
Changes following review

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

ghalliday committed:
Add initial test script

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

ghalliday committed:
Add code to sanity check the UDP settings

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

ghalliday committed:
Option to configure the number of free slots required to grant a permit

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

ghalliday commented:
@richardkchapman Closing to open a new PR without all the cleaned up comments.  Probably helpful if you read through this PR to check I have addressed all the comments though.
